include/oneapi/dpl/pstl/onedpl_config.h

```diff
@@ -301,6 +301,12 @@
 #define _ONEDPL_CPP20_SHIFT_LEFT_RIGHT_PRESENT                                                                         \
     (_ONEDPL___cplusplus >= 202002L && (_MSC_VER >= 1921 || _GLIBCXX_RELEASE >= 10))
 
+#if _ONEDPL___cplusplus >= 202002L
+#   define _ONEDPL_CPP20_RANGES_PRESENT (__cpp_lib_ranges >= 201911L)
```

Comment:
I don't want to overcomplicate this, but maybe we should move this into the `#if _ONEDPL_STD_FEATURE_MACROS_PRESENT` block directly below similar to what we've done for the concepts feature macros? If feature macros aren't present, then this may be problematic with how its currently written.

Also, since we are now using concepts should we have some check of `_ONEDPL_CPP20_CONCEPTS_PRESENT` somewhere which checks for both compiler and library support or is it a safe assumption that this will be present?

Reply:
Definition of `_ONEDPL_STD_FEATURE_MACROS_PRESENT` is depend on `__has_include(<version>)`...
Does it mean that `<version>` file might not be present on a system?

Reply:
`https://en.cppreference.com/w/cpp/header/version` is codified in C++20 that defines the feature macros, so it is present on Windows. The check for `__has_include(<version>)` is just to make sure that this C++20 feature is implemented in the implementation the user is using.

Reply:
I suppose Matthew is talking about a case like [that](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxBIArKQADqgKhE4MHt6%2BekkpjgJBIeEsUTFc8XaYDmlCBEzEBBk%2Bfly2mPZ5DDV1BAVhkdFxtrX1jVktCsM9wX3FA2UAlLaoXsTI7BzmAMzByN5YANQmm25OE8SYrEfYJhoAgjf3ZttU%2BwD6r8gJ3gpfXgqHm2umwAIvszBpwRCAg8tvs4VgaCF9m5lMpwa8AEq3UIAcWwQleygx%2BOwoQAKvsIO9PglXrQ8BFXsRDMBMP8rkdQeCuABOLhcALzGFPNpKYWbOH7BHTZGo9FY3H4wnEoSkikacWuPBUGF3YIEfYsJjBCBCu4mADsVjukv2E3QIBQywNRzcrtlaI0mOxeIJRJJ5IBbuOdoIDpArloR2tt1t9sdaC8LuO7upvx%2B3yD7vjEYY6CjmxjtvOBBWDH2GsLMItwIeHEWtE4sV4fm4vFQnDdlmsduWq0whyePFIBE09cWAGsQJILQA6ABs882fIAHJtNrEzDz55JNvpOJJeCwQLENKRW1pSB2OLwFCAz6OOFpFnBYDBEE6WAk6NFyJQ0F%2BP4xLsLJcBoGgtDQtAENEd4QBEY6kBEwR1AAnpww7IcwxCoQA8hE2iVI%2Bw4AWwgi4QwtDoU%2BvBYBEXjAG4Yi0HebakFgRpGOINHsXg5xVAAbmyiGYKolRJusl76m0iH0hEzI4R4WCIQQxB4MebFCcQETJJgwKYJxwD0kYY6LFQBjAAoABqeCYAA7rhCSMBhvD8IIIhiOwUgyIIigqOoPG6C0BgmaY3aWPoDJ3pAiyoAkHSsQAtPanJhVYljgu2WlqVg0Wmq07RpC4eajM0pCBNMRQlNkySpAIpU1bkaS9FVcwFUR1STA14xtB1AhdPULX9KUQzdN1o2DZVw0SIsCh9msM37hwzbnoh177KoK7zolO77CBRj7GBs4aLOXCUrghAkIOmxcPMvCPs%2BiwIBcWAxPlU6bHOkgrrEN1cMua6xN9K5LYepDHuus7rtDMOw/Oq08det73iOpmkK%2BH7OgkSZ/hAAHfvQxChKw6ybdtu3AMgyCUqpXgMBOd3lfgRA5XobnCKI4jeezflqIhQWkHZzIJC5DZNi2a2cLhSbYwaqAvGTO2SHtIXAIdJ0nWdEAeIBhPXbd92mU9L0DPljYcGDEObFDsO2%2Bu8MXu2nDIw%2BaMYxASBYAJeBrLj%2BNAcTbCcIrFNU4dkinbRmDewtCiobsyXfsQO6JXTE4MKgdkMKn9MZ1nTOXazkGyB5XPSDzSh84FegVFUzgQK440VYU00tDkdXpJ4TSNR3Q2zCNtcdANDRd2M7V150kx99V4xdaPZUTN009zIsqmYJgeAOhoYvLRLiOcPpMcDjZ9nRBtW1KyrLLq8dp3nczV1bC0yKoLrZ9P4zD3jqDR4gJDdt2wRpeJGtgUZf3mJOE8Zgbbzg0CuLgi4ZwaA3BaJamw97AOdqjGiEClpmF/mBM8jsrxYPAYsLSKRnCSCAA%3D%3D) (clang 10). 
It can be fixed by adding `defined(__cpp_lib_ranges)` condition.

Reply:
Yes, I'm referring to issues such as that. I think some libraries defined a `version` header prior to C++20, so what we are currently doing may not be robust enough.

Dmitriy's suggestion would resolve this issue. We should probably consolidate how we do this for the existing feature macro checks but that can be done in a separate PR.